iroha_introspect/src/lib.rs

```diff
@@ -0,0 +1,358 @@
+        Self::introspect(&mut map);
+        map
+    }
+
+    /// Introspect function. Give it empty map and it will return description of types
+    /// related to this type
+    fn introspect(metamap: &mut MetaMap);
+}
+
+/// Metadata
+#[derive(Debug, Eq, PartialEq, Hash, Ord, PartialOrd, Serialize, Clone)]
+pub enum Metadata {
+    /// Structure with named fields
+    Struct(NamedFieldsMeta),
+    /// Unnamed structure
+    TupleStruct(UnnamedFieldsMeta),
+    /// Enum
+    Enum(EnumMeta),
+    /// Integer
+    Int(IntMode),
+    /// String
+    String,
+    /// Bool
+    Bool,
+    /// Array
+    Array(ArrayMeta),
+    /// Vector with type
+    Vec(String),
+    /// Option with type
+    Option(String),
+    /// Map
+    Map(MapMeta),
+    /// Result
+    Result(ResultMeta),
+}
+
+/// Array metadata
+#[derive(Debug, Eq, PartialEq, Hash, Ord, PartialOrd, Serialize, Clone)]
+pub struct ArrayMeta {
+    ty: String,
+    len: usize,
+}
+
+/// Named fields
+#[derive(Debug, Eq, PartialEq, Hash, Ord, PartialOrd, Serialize, Clone)]
+pub struct NamedFieldsMeta {
+    /// Fields
+    pub declarations: Vec<Declaration>,
+    //todo add collection of properties meta defined in struct
+}
+
+/// Field
+#[derive(Debug, Eq, PartialEq, Hash, Ord, PartialOrd, Serialize, Clone)]
+pub struct Declaration {
+    /// Field name
+    pub name: String,
+    /// Type
+    pub ty: String,
+}
+
+/// Unnamed fileds
+#[derive(Debug, Eq, PartialEq, Hash, Ord, PartialOrd, Serialize, Clone)]
+pub struct UnnamedFieldsMeta {
+    /// Field types
+    pub types: Vec<String>,
+    //todo add collection of properties meta defined in struct
+}
+
+/// Enum metadata
+#[derive(Debug, Eq, PartialEq, Hash, Ord, PartialOrd, Serialize, Clone)]
+pub struct EnumMeta {
+    /// Enum variants
+    pub variants: Vec<EnumVariant>,
+}
+
+/// Enum variant
+#[derive(Debug, Eq, PartialEq, Hash, Ord, PartialOrd, Serialize, Clone)]
+pub struct EnumVariant {
+    /// Enum variant name
+    pub name: String,
+    /// Its discriminant (or identifier)
+    pub discriminant: u8,
+    /// Its type
+    pub ty: Option<String>,
+    //todo add collection of properties meta defined in enum variant
+}
+
+/// Result variant
+#[derive(Debug, Eq, PartialEq, Hash, Ord, PartialOrd, Serialize, Clone)]
+pub struct ResultMeta {
+    /// Ok type
+    pub ok: String,
+    /// Err type
+    pub err: String,
+}
+
+/// Map variant
+#[derive(Debug, Eq, PartialEq, Hash, Ord, PartialOrd, Serialize, Clone)]
+pub struct MapMeta {
+    /// Key type
+    pub key: String,
+    /// Value type
+    pub value: String,
+}
+
+/// Integer mode
+#[derive(Debug, Eq, PartialEq, Hash, Ord, PartialOrd, Serialize, Clone, Copy)]
+pub enum IntMode {
+    /// Fixed width
+    FixedWidth,
+    /// Scale compact
+    Compact,
+}
+
+/// Compact predicate. Just for documentation purposes
+#[derive(Debug, Eq, PartialEq, Hash, Ord, PartialOrd, Serialize, Clone, Copy)]
+pub struct Compact<T>(T);
+
+macro_rules! impl_introspect_i {
```

Comment:
```suggestion
macro_rules! impl_introspect_integer {
```

Reply:
It is clear from context, and it is used just once. i32, i64 is an ok naming to me, as the `impl_introspect_i!`

Reply:
But there are also `u` and it is our policy to use full names even in macros :)